auction/src/auction.ts

```diff
@@ -0,0 +1,101 @@
+export type Buyer = {
+    demand: number;
+    price: number;
+    account: string;
+}
+
+export type Seller = {
+    supply: number;
+    account: string;
+}
+
+export type Transaction = {
+    sellerKey: string;
+    buyerKey: string;
+    energy: number;
+    price: number;
+}
+
+export class Auction {
+    private buyers: Buyer[];
+
+    private sellers: Seller[];
+    
+    public constructor(sellers: Seller[], buyers: Buyer[]){
+      if (this._inputValid(sellers, buyers)) {
+          this.sellers = sellers;
+          this.buyers = buyers;
+      } else { 
+          this.sellers = [];
+          this.buyers = [];
+      }
+
+      this.buyers.sort((buyer1, buyer2) => buyer2.price - buyer1.price); // Highest bid wins
+      this.sellers.sort((seller1, seller2) => seller2.supply - seller1.supply); // Highest supply gets to sell first (re-sorted after each selling part of supply)
+    }
+
+    public auction(): Transaction[] {
+        let transactions: Transaction[] = [];
+        
+        // Main part of the auction - we iteratively serve the highest bidder
+        while (1 < this.buyers.length && 0 < this.sellers.length) { // Need to handle corner cases
+            let bid = this.buyers[0];
+            transactions = transactions.concat(this._serveWinner(bid, this.buyers[1].price)); //if we have atleast 2 buyers
+            
+            if (bid.demand == 0)
+                this.buyers.splice(0, 1);
+        }
+
+        // Special case - we have excess supply and exactly one buyer to serve.
+        if (this.buyers.length == 1 && 0 < this.sellers.length) {
+            transactions = transactions.concat(this._serveWinner(this.buyers[0], this.buyers[0].price));
+
+            if (this.buyers[0].demand == 0) // Only remove the last bidder, if their entire demand has been met.
+                this.buyers.splice(0,1);
+        }
+        
+        this.sellers.map((seller) => console.log(seller.account,seller.supply)); // Log sellers with excess supply after the auction
+        this.buyers.map((buyer) => console.log(buyer.account, buyer.demand)); // Log bidders with un-met demand
+
+        this._handleTransactions(transactions);
+
+        return transactions;
+    }
+    
+    private _inputValid(sellers: Seller[], buyers: Buyer[]): boolean {
+        const sellerBool: boolean = sellers.every((seller) => seller.supply > 0 && Number.isInteger(seller.supply));
```

Comment:
```suggestion
        const sellerBool = sellers.every((seller) => seller.supply > 0 && Number.isInteger(seller.supply));
```

Bare lad typescript infer imo, det er den rigtig god til :)